<<<<<<< HEAD
library(data.table)
library(dplyr)
library(coloc)
library(hash)
library(optparse)
library(locuscomparer)

# takes the binary files for the gwas and e/pqtl coloc objects as inputs for the Rscript #
option_list <- list(
  make_option("--gwas", type = "character", help = "Filepath containing the binary file of the GWAS coloc object"),
  make_option("--twas", type = "character", help = "Filepath containing the binary file of the e/pQTL coloc object"),
  make_option("--ref_dir", type = "character", help = "Directory contating the outputs of the full genome analysis"),
  make_option("--gene_dir", type = "character", help = "Directory containing the gene of interest")
)

opt <- parse_args(OptionParser(option_list = option_list))

gwascoloc <- readRDS(opt$gwas)
twascoloc <- readRDS(opt$twas)
gene.dir <- opt$gene_dir
ref.dir <- opt$ref_dir

if(!dir.exists(paste0(gene.dir, "mv_results/"))){
  dir.create(paste0(gene.dir, "mv_results/"), recursive = TRUE)
}
res.dir <- paste0(gene.dir, "mv_results/")


#make a list of EUR individuals in 1000 Genomes
for(i in list.files(ref.dir)){
  if(grepl(".psam", i) == TRUE){
    pops = fread(paste0(ref.dir, i))
  }
}
for(i in list.files(ref.dir)){
  if(grepl(".fam", i) == TRUE){
    fams = fread(paste0(ref.dir, i))
  }
}
pops <- as.data.frame(pops)
eur = filter(pops,SuperPop=="EUR")
#note the FID column (column 1) in /homes/data/1000g_ref_data/all_phase3.rsid.maf001.fam is all zeros
eurlist = dplyr::select(eur, `#IID`)
# adds a column with zeroes for FID and then makes a new data.table with only the FID and #IID columns #


fwrite(eurlist,paste0(res.dir, 'id_list'),col.names=FALSE, sep="\t")
# outputs the results so they can be read by plink later #

#write list of SNPs in coloc df's
fwrite(data.frame(twascoloc$snp), file = paste0(res.dir,"snp_list"), col.names = FALSE)
# outputs the names of the snps that made it through coloc such that it can be read by plink #

save.image('./test.RData')
setwd(res.dir)
system(paste0("/home/2025/cdotson/plink --bfile ", ref.dir, strsplit(list.files(ref.dir)[1], '.bed'), " --extract ", paste0(res.dir, 'snp_list'), " --id-delim ':' --keep ", paste0(res.dir, 'id_list'), " --maf 0.01 --recode A --make-just-bim --out ./coloc_region --allow-extra-chr"))

#read the EUR .raw file generated by plink into R.#
geno = fread(paste0(res.dir,"coloc_region.raw"))
# reads the raw files and tells you if the individual is either a homozygote for the reference (0), heterozygous (1), and homozygote for the variant (2) #

# we want to put genotypes in NxP matrix (N=people, P=snps) with no other columns
genomat = as.matrix(geno[,7:length(geno)])
# filter out all the columns that don't have snp information #

#now filter the gwascoloc and pqtlcoloc df's to those SNPs that were in EUR (in genomat)
#get a list of SNPs in genomat
snplist = colnames(genomat)

#remove the last 2 characters (_N) from genomat colnames to match topsleep with substr()
snplist = substr(snplist, 1, nchar(snplist)-2)
# removed the last 2 characters from the string #

#rename col names of R to match coloc df's
colnames(genomat) = snplist

#check pQTL and 1000G EUR alleles (ask, what is coded as 1?), if need to, flip BETA signs.
#col 5 in bim is A1 (assigned 1 in dosage raw file)
bim = fread("./plink_results/EUR_coloc-region.bim")

susiesnpsbim = inner_join(matchsnps, bim, by=c("ID"="V2"))
# make a data.table that joins the rows of matchsnps and bim by ID and V2, repsectively

#pull SNPs that  match (assumes C|G and A|T SNPs aren't flipped)
matchbim = susiesnpsbim[(susiesnpsbim$A1.x==susiesnpsbim$V5 & susiesnpsbim$REF == susiesnpsbim$V6),]
# same as we did before except with the combined snps determined for susie @
#remove ambiguous strand SNPs (A/T or C/G)

b = susiesnpsbim[!(susiesnpsbim$REF=="A" & susiesnpsbim$ALT=="T") & !(susiesnpsbim$REF=="T" & susiesnpsbim$ALT=="A") & !(susiesnpsbim$REF=="C" & susiesnpsbim$ALT=="G") & !(susiesnpsbim$REF=="G" & susiesnpsbim$ALT=="C") ]
# same as before #

#of non-ambiguous, pull SNPs that are flipped (or the complement bases match or are flipped)
compmatchbim = b[(b$A1.x==values(bases,keys=b$V5) & b$REF == values(bases,keys=b$V6)),]
# same as before #

flippedbim = b[(b$A1.x==b$V6 & b$REF == b$V5),]
# same as before #
compflippedbim = b[(b$A1.x==values(bases,keys=b$V6) & b$REF == values(bases,keys=b$V5)),]
# same as before #

#if flipped, change sign of cancer and pQTL beta, check to see if any SNPs are in flipped df's with if stmt
if(dim(flippedbim)[1] > 0){
  flippedbim = mutate(flippedbim,BETA.x = -1*BETA.x,BETA.y = -1*BETA.y)
}
if(dim(compflippedbim)[1] > 0){
  compflippedbim = mutate(compflippedbim,BETA.x = -1*BETA.x,BETA.y = -1*BETA.y)
}
# same as before #

#bind all and sort by position
matchbimsnps = rbind(matchbim, compmatchbim, flippedbim, compflippedbim) %>% arrange(POS)

#update snplist
snplist = matchbimsnps$ID
#filter genomat to snps in snplist
x = genomat[,colnames(genomat) %in% snplist]
#calculate the correlation matrix of the genotypes, this is needed for susie
R = cor(x)

#filter the matchbimsnps df to just the snps in snplist with dplyr
susiesnps = filter(matchbimsnps, ID %in% snplist)
#add LD to filtered coloc df's
gwascolocsusie = list("beta" = susiesnps$BETA.y, "varbeta" = (susiesnps$SE.y)^2, "snp" = susiesnps$ID, "position" = susiesnps$POS,
                      "type" = "cc", "LD"=R, "N" = 100000) # need to add N, check ref later
pqtlcolocsusie = list("beta" = setNames(susiesnps$BETA.x, susiesnps$ID), "varbeta" = setNames(susiesnps$SE.x^2, susiesnps$ID), "snp" = susiesnps$ID,
                      "position" = susiesnps$POS,"type" = "quant", "N" = susiesnps$OBS_CT[1], "MAF" = susiesnps$A1_FREQ, "sdY"=1, "LD"=R)

check_dataset(gwascolocsusie,req="LD")
# if NULL, that means the correlation matrix was added to the object #
check_dataset(pqtlcolocsusie,req="LD")

plot_dataset(gwascolocsusie)

plot_dataset(pqtlcolocsusie)

sgwas = runsusie(gwascolocsusie)
# runs a Sum of Single Effects (SuSiE) Regression 

spqtl = runsusie(pqtlcolocsusie)

summary(sgwas)

summary(spqtl)

susie.res1=coloc.susie(sgwas, spqtl)
# actually performs the colocalization #

print(susie.res1$summary)

sigrows = which(susie.res1$summary$PP.H4.abf > 0.5)
sigrows

sensitivity(susie.res,"H4 > 0.5",row=1,dataset1=gwascolocsusie,dataset2=pqtlcolocsusie)

for(i in sigrows){
  sensitivity(susie.res,"H4 > 0.5",row=i,dataset1=gwascolocsusie,dataset2=pqtlcolocsusie)
}

sgwas = runsusie(gwascolocsusie, coverage=0.1)

spqtl = runsusie(pqtlcolocsusie, coverage=0.1)

summary(sgwas)

summary(spqtl)

susie.res2=coloc.susie(sgwas, spqtl)

print(susie.res2$summary)

sensitivity(susie.res2,"H4 > 0.5",row=1,dataset1=gwascolocsusie,dataset2=pqtlcolocsusie)

for(i in sigrows){
  sensitivity(susie.res,"H4 > 0.5",row=i,dataset1=gwascolocsusie,dataset2=pqtlcolocsusie)
}

matchbimsnps.df <- as.data.frame(matchbimsnps)
gwas.snps <- matchbimsnps.df[,c('ID','P.x')]
colnames(gwas.snps) <- c('rsid', 'pval')

eqtl.snps <- matchbimsnps.df[,c('ID','P.y')]
colnames(eqtl.snps) <- c('rsid', 'pval')

locuscompare(in_fn1 = gwas.snps, in_fn2 = eqtl.snps, title1 = 'eQTL', title2 = 'GWAS')
=======
library(data.table)
library(dplyr)
library(coloc)
library(hash)
library(optparse)
library(locuscomparer)

# takes the binary files for the gwas and e/pqtl coloc objects as inputs for the Rscript #
option_list <- list(
  make_option("--gwas", type = "character", help = "Filepath containing the binary file of the GWAS coloc object"),
  make_option("--twas", type = "character", help = "Filepath containing the binary file of the e/pQTL coloc object"),
  make_option("--ref_dir", type = "character", help = "Directory contating the outputs of the full genome analysis"),
  make_option("--gene_dir", type = "character", help = "Directory containing the gene of interest")
)

opt <- parse_args(OptionParser(option_list = option_list))

gwascoloc <- readRDS(opt$gwas)
twascoloc <- readRDS(opt$twas)
gene.dir <- opt$gene_dir
ref.dir <- opt$ref_dir

if(!dir.exists(paste0(gene.dir, "mv_results/"))){
  dir.create(paste0(gene.dir, "mv_results/"), recursive = TRUE)
}
res.dir <- paste0(gene.dir, "mv_results/")


#make a list of EUR individuals in 1000 Genomes
for(i in list.files(ref.dir)){
  if(grepl(".psam", i) == TRUE){
    pops = fread(paste0(ref.dir, i))
  }
}
for(i in list.files(ref.dir)){
  if(grepl(".fam", i) == TRUE){
    fams = fread(paste0(ref.dir, i))
  }
}
pops <- as.data.frame(pops)
eur = filter(pops,SuperPop=="EUR")
#note the FID column (column 1) in /homes/data/1000g_ref_data/all_phase3.rsid.maf001.fam is all zeros
eurlist = dplyr::select(eur, `#IID`)
# adds a column with zeroes for FID and then makes a new data.table with only the FID and #IID columns #


fwrite(eurlist,paste0(res.dir, 'id_list'),col.names=FALSE, sep="\t")
# outputs the results so they can be read by plink later #

#write list of SNPs in coloc df's
fwrite(data.frame(twascoloc$snp), file = paste0(res.dir,"snp_list"), col.names = FALSE)
# outputs the names of the snps that made it through coloc such that it can be read by plink #

save.image('./test.RData')
setwd(res.dir)
system(paste0("/home/2025/cdotson/plink --bfile ", ref.dir, strsplit(list.files(ref.dir)[1], '.bed'), " --extract ", paste0(res.dir, 'snp_list'), " --id-delim ':' --keep ", paste0(res.dir, 'id_list'), " --maf 0.01 --recode A --make-just-bim --out ./coloc_region --allow-extra-chr"))

#read the EUR .raw file generated by plink into R.#
geno = fread(paste0(res.dir,"coloc_region.raw"))
# reads the raw files and tells you if the individual is either a homozygote for the reference (0), heterozygous (1), and homozygote for the variant (2) #

# we want to put genotypes in NxP matrix (N=people, P=snps) with no other columns
genomat = as.matrix(geno[,7:length(geno)])
# filter out all the columns that don't have snp information #

#now filter the gwascoloc and pqtlcoloc df's to those SNPs that were in EUR (in genomat)
#get a list of SNPs in genomat
snplist = colnames(genomat)

#remove the last 2 characters (_N) from genomat colnames to match topsleep with substr()
snplist = substr(snplist, 1, nchar(snplist)-2)
# removed the last 2 characters from the string #

#rename col names of R to match coloc df's
colnames(genomat) = snplist

#check pQTL and 1000G EUR alleles (ask, what is coded as 1?), if need to, flip BETA signs.
#col 5 in bim is A1 (assigned 1 in dosage raw file)
bim = fread("./plink_results/EUR_coloc-region.bim")

susiesnpsbim = inner_join(matchsnps, bim, by=c("ID"="V2"))
# make a data.table that joins the rows of matchsnps and bim by ID and V2, repsectively

#pull SNPs that  match (assumes C|G and A|T SNPs aren't flipped)
matchbim = susiesnpsbim[(susiesnpsbim$A1.x==susiesnpsbim$V5 & susiesnpsbim$REF == susiesnpsbim$V6),]
# same as we did before except with the combined snps determined for susie @
#remove ambiguous strand SNPs (A/T or C/G)

b = susiesnpsbim[!(susiesnpsbim$REF=="A" & susiesnpsbim$ALT=="T") & !(susiesnpsbim$REF=="T" & susiesnpsbim$ALT=="A") & !(susiesnpsbim$REF=="C" & susiesnpsbim$ALT=="G") & !(susiesnpsbim$REF=="G" & susiesnpsbim$ALT=="C") ]
# same as before #

#of non-ambiguous, pull SNPs that are flipped (or the complement bases match or are flipped)
compmatchbim = b[(b$A1.x==values(bases,keys=b$V5) & b$REF == values(bases,keys=b$V6)),]
# same as before #

flippedbim = b[(b$A1.x==b$V6 & b$REF == b$V5),]
# same as before #
compflippedbim = b[(b$A1.x==values(bases,keys=b$V6) & b$REF == values(bases,keys=b$V5)),]
# same as before #

#if flipped, change sign of cancer and pQTL beta, check to see if any SNPs are in flipped df's with if stmt
if(dim(flippedbim)[1] > 0){
  flippedbim = mutate(flippedbim,BETA.x = -1*BETA.x,BETA.y = -1*BETA.y)
}
if(dim(compflippedbim)[1] > 0){
  compflippedbim = mutate(compflippedbim,BETA.x = -1*BETA.x,BETA.y = -1*BETA.y)
}
# same as before #

#bind all and sort by position
matchbimsnps = rbind(matchbim, compmatchbim, flippedbim, compflippedbim) %>% arrange(POS)

#update snplist
snplist = matchbimsnps$ID
#filter genomat to snps in snplist
x = genomat[,colnames(genomat) %in% snplist]
#calculate the correlation matrix of the genotypes, this is needed for susie
R = cor(x)

#filter the matchbimsnps df to just the snps in snplist with dplyr
susiesnps = filter(matchbimsnps, ID %in% snplist)
#add LD to filtered coloc df's
gwascolocsusie = list("beta" = susiesnps$BETA.y, "varbeta" = (susiesnps$SE.y)^2, "snp" = susiesnps$ID, "position" = susiesnps$POS,
                      "type" = "cc", "LD"=R, "N" = 100000) # need to add N, check ref later
pqtlcolocsusie = list("beta" = setNames(susiesnps$BETA.x, susiesnps$ID), "varbeta" = setNames(susiesnps$SE.x^2, susiesnps$ID), "snp" = susiesnps$ID,
                      "position" = susiesnps$POS,"type" = "quant", "N" = susiesnps$OBS_CT[1], "MAF" = susiesnps$A1_FREQ, "sdY"=1, "LD"=R)

check_dataset(gwascolocsusie,req="LD")
# if NULL, that means the correlation matrix was added to the object #
check_dataset(pqtlcolocsusie,req="LD")

plot_dataset(gwascolocsusie)

plot_dataset(pqtlcolocsusie)

sgwas = runsusie(gwascolocsusie)
# runs a Sum of Single Effects (SuSiE) Regression 

spqtl = runsusie(pqtlcolocsusie)

summary(sgwas)

summary(spqtl)

susie.res1=coloc.susie(sgwas, spqtl)
# actually performs the colocalization #

print(susie.res1$summary)

sigrows = which(susie.res1$summary$PP.H4.abf > 0.5)
sigrows

sensitivity(susie.res,"H4 > 0.5",row=1,dataset1=gwascolocsusie,dataset2=pqtlcolocsusie)

for(i in sigrows){
  sensitivity(susie.res,"H4 > 0.5",row=i,dataset1=gwascolocsusie,dataset2=pqtlcolocsusie)
}

sgwas = runsusie(gwascolocsusie, coverage=0.1)

spqtl = runsusie(pqtlcolocsusie, coverage=0.1)

summary(sgwas)

summary(spqtl)

susie.res2=coloc.susie(sgwas, spqtl)

print(susie.res2$summary)

sensitivity(susie.res2,"H4 > 0.5",row=1,dataset1=gwascolocsusie,dataset2=pqtlcolocsusie)

for(i in sigrows){
  sensitivity(susie.res,"H4 > 0.5",row=i,dataset1=gwascolocsusie,dataset2=pqtlcolocsusie)
}

matchbimsnps.df <- as.data.frame(matchbimsnps)
gwas.snps <- matchbimsnps.df[,c('ID','P.x')]
colnames(gwas.snps) <- c('rsid', 'pval')

eqtl.snps <- matchbimsnps.df[,c('ID','P.y')]
colnames(eqtl.snps) <- c('rsid', 'pval')

locuscompare(in_fn1 = gwas.snps, in_fn2 = eqtl.snps, title1 = 'eQTL', title2 = 'GWAS')
>>>>>>> e257f1465ff46f9641f0e48d2f0633777cd773a5
