---
title: "Visualization of coloc Results"
output: html_document
params: 
  rdata_path: ""
---


```{r setup, include=FALSE, warning=FALSE}
setwd('/home/2025/cdotson/test')
load(params$rdata_path)
library(locuscomparer)
```

# `r gene.name`

## locuscompareR Preliminary Analysis

```{r print, echo = FALSE, warning=FALSE,message=FALSE,error=FALSE}
locuscompare(in_fn1 = gwassnps.df, in_fn2 = twassnps.df, title1 = 'GWAS', title2 = 'eQTL', population = super_pop, genome = 'hg38')
```

Interpretation: If the causal genetic variant for higher rates of the two traits can be detected, then we would expect it to be exactly or linked with the SNP `r lead.snp`, which is located on Chromosome `r leadsnp.loc$chr[1]` at position `r leadsnp.loc$pos[1]`

### SUSIE Run with GWAS Data
There were `r gwas.count` SNP(s) across `r gwas_cc.count` credible set(s) detected in the GWAS data.

### SUSIE Run with TWAS Data 
There were `r twas.count` SNP(s) across `r twas_cc.count` credible set(s) detected in the TWAS data. 
